### PR TITLE
Add clustering example and modify CBF dataset function

### PR DESCRIPTION
**Pull request by lucasplagwitz**
Hello,

because pyts comes with all the necessary requirements for time series clustering, I set up an example regarding different metrics in this context. Even though pyts focuses on classification, I think a concrete example of using BOSS in an unsupervised way is quite useful. What do you think?

The example refers to the original BOSS paper. Since a slightly different form of the data set was used there, I shifted each time series in a second attempt.

### CODE REVIEW

**Comment by lucasplagwitz** [comment]
Thank you for the quick reply. I agree, simply "shifting" is really not the best idea. In fact, I've thought about the noise as well, but find the aspect between global and local distinction a bit more appealing. If you are familiar with the BOSS paper it's obvious, it's true, but this example is meant more for those who haven't read it yet/want to recreate it.

I just looked again at the reference of the CBF dataset, and there the limits a and b are set per sample. The pyts implementation sets them once per dataset, if I see it correctly. Is this intentional?
By simply modifying [these lines](https://github.com/johannfaouzi/pyts/blob/28742ea713b19699ada13600bf6e10a11bc1b192/pyts/datasets/make.py#L132) to:
```python
    a = rng.randint(16, 33, size=(n_samples, 1))
    b = rng.randint(32, 97, size=(n_samples, 1)) + a
```
and making a few small changes, we get a comparable dataset. Could also be activated by a parameter. 

If you then add the noise in the second line, you get a result like this:
boss_clustering
It must be said, however, that the results do not seem entirely constant.

**Comment by johannfaouzi** [comment]
To be honest, it was not clear to me if the parameters should be the same or not, so I made this choice of making them all identical for a given dataset. With more hindsight, this approach seems less natural. I just checked the original CBF dataset and the parameters `a`, `b` and `eta` are obviously different for different samples. Thank you for pointing this out!

Gaussian noise from a standard normal distribution is added with the current implementation (and I think that it is necessary to have some noise, otherwise it's too easy), but the value of the standard deviation could be added as an optional parameter (default value = 1.0). However, I'm not sure to understand if you added Gaussian noise only once or twice (because there's already Gaussian noise in the current implementation):

https://github.com/johannfaouzi/pyts/blob/28742ea713b19699ada13600bf6e10a11bc1b192/pyts/datasets/make.py#L135-L143

Obviously the more noise you add, the harder it becomes.

On another note, I wonder if it would not be better to specify the number of clusters in this case. I know that this information is not available in practice, but in this case we don't really want to assess the capability of the clustering algorithm to identify the right number of clusters, and we are more focused on the metrics used to compare samples.

**Comment by lucasplagwitz** [comment]
I have now updated the CBF dataset. Please let me know if this should not be adjusted for now.

I agree with you on both points. My idea was really just to put more noise over it. That's not the prettiest approach though. With the fixed cluster count, there is the problem of visualization: The dendrogram is only usable if you have calculated the complete tree, right? I can't think of a better visualization at this point.

However, I also find a measurement of the performance very useful. Can you take a look at my updated example? According to this, BOSS would be significantly worse than DTW. Can you find a mistake? Is it due to certain BOSS parameters?

The text will be adjusted later, when the results of the example have been determined.

Thanks a lot for your help,
Lucas

**Comment by johannfaouzi** [comment]
> I have now updated the CBF dataset. Please let me know if this should not be adjusted for now.

No, this is fine, thanks very much! However, you have to update the tests for this function because these parameters are now instances of the `np.ndarray` class:

https://github.com/johannfaouzi/pyts/blob/91b70388b4a036625fdb7bacc71150bc983dc397/pyts/datasets/tests/test_make.py#L68-L70

> I agree with you on both points. My idea was really just to put more noise over it. That's not the prettiest approach though. With the fixed cluster count, there is the problem of visualization: The dendrogram is only usable if you have calculated the complete tree, right? I can't think of a better visualization at this point.

You can still compute the whole tree using `compute_full_tree=True`. I thought that setting the number of clusters to 3 would limit the number of colors in the dendrogram, but it does not seem to be the case.

I think that plotting a single clustering run, but computing the average metric over several runs, is not very intuitive (the displayed performance does not correspond to the plot).  

> However, I also find a measurement of the performance very useful. Can you take a look at my updated example? According to this, BOSS would be significantly worse than DTW. Can you find a mistake? Is it due to certain BOSS parameters?

I did not find any mistake. However, the BOSS algorithm has many hyper-parameters, so it's hard to know which values of these hyper-parameters are optimal. If you just pick the first 6 time series in the original CBF dataset, you obtain a similar figure to the one in the paper. From here on out, there may be several reasons that it does not generalize to other datasets (wrong implementation, suboptimal values of hyper-parameters, cherry-picked example, etc.). It's really hard to tell and it unfortunately may take a huge amount of time to find out.

In this particular case (CBF dataset), I find the amount of noise substantial and the true signal is really simple (a polynomial of degree at most one), so I'm not sure FFT is that useful to identify such signals. In the BOSS paper they use 1NN classification, so basically, for testing, you need only one training example that is very similar. For clustering, you take into account all the training samples (with their heterogeneity), so it's much harder.

What makes you think that BOSS should be much better than DTW? Is it based on the BOSS paper?

**Comment by lucasplagwitz** [comment]
Very helpful comments, thank you!

```
What makes you think that BOSS should be much better than DTW? Is it based on the BOSS paper?
```
In fact, it's mostly the BOSS paper. I didn't read it sensibly and didn't separate Figure 2 clearly enough from Figure 1 (classification vs. clustering). Anyway, Figure 1 suggests an advantage of clustering over DTW, but just only for the 6 samples. The CBF dataset really isn't the best example to demonstrate BOSS-based clustering. Perhaps at this point there would be space for a second example that contrasts different structures of time series.

Anyway, I think the adapted example is helpful for newcomers to time series clustering. Do you notice any other points that should be adjusted?

File: pyts/datasets/make.py [hunk 1]
```diff
@@ -129,16 +129,17 @@ def make_cylinder_bell_funnel(
...
-    eta = rng.randn()
+    a = rng.randint(16, 33, size=(n_samples, 1))
+    b = rng.randint(32, 97, size=(n_samples, 1)) + a
+    eta = rng.randn(n_samples, 1)
     epsilon = rng.randn(n_samples, 128)
 
     arange = np.tile(np.arange(1, 129), n_samples).reshape(n_samples, 128)
-    additional_term = (6 + eta) * np.logical_and(arange >= a, arange <= b)
+    additional_term = (6 + eta) * np.logical_and(arange >= np.repeat(a, 128, axis=1),
+                                                 arange <= np.repeat(b, 128, axis=1))
```
Comment (johannfaouzi) [inline pyts/datasets/make.py:139]: Lines too long

File: examples/clustering/plot_dtw_boss.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+suitable metric should be chosen: While the Euclidean distance and DTW are
+sensitive to time-dependent events, clustering with BOSS does not consider the
+temporal component outside the sliding window.
+
+References
+----------
+
+[1] P. Schäfer, "The BOSS is concerned with time series classification
+    in the presence of noise". Data Mining and Knowledge Discovery,
+    29(6), 1505-1530 (2015).
```
Comment (johannfaouzi) [inline examples/clustering/plot_dtw_boss.py:23]: It does not display well. I think that it takes the `P.` as a symbol of enumeration (similar to an enumerate environment in LaTeX). I had this issue with `L. Breiman`, and replacing it with `Leo Breiman` fixed it. You may try using the full first name (Patrick) to fix it.

File: examples/clustering/plot_dtw_boss.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    model = AgglomerativeClustering(compute_full_tree=True,
+                                    compute_distances=True,
+                                    n_clusters=3, affinity="precomputed",
+                                    linkage="complete")
+    cluster = model.fit_predict(dist_mat)
+    score = round(homogeneity_score(labels_true=y, labels_pred=cluster), 2)
+
+    plot_dendrogram(model, orientation='left',
+                    ax=axes[k_axis], labels=list(zip(y, cluster)))
```
Comment (johannfaouzi) [inline examples/clustering/plot_dtw_boss.py:99]: Cluster labels are not meaningful (any permutation of the labels corresponds to the same clustering), so I'm not sure that it is relevant to display the raw cluster labels. If you really want to display them, I think that you have to find the best permutation of the cluster labels (the ones that maximizes the accuracy).

File: examples/clustering/plot_dtw_boss.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+=========================================
+Time Series Clustering with DTW and BOSS
+=========================================
+
+This example shows the differences between various metrics
+related to time series clustering. Besides the Euclidean distance,
+:func:`pyts.metrics.dtw` and :func:`pyts.metrics.boss` are considered to
+analyze the :func:`pyts.datasets.make_cylinder_bell_funnel` dataset.
+In contrast to reference [1], DTW-based clustering shows the best results
+here - even when repeated over multiple seeds. The reason for this could be
```
Comment (johannfaouzi) [inline examples/clustering/plot_dtw_boss.py:11]: I don't think that you repeat the experiments with different random seeds now. Even if you tried it out, I'm not sure that it should be mentioned in the documentation of this example if it does not illustrate it.

**Review by johannfaouzi — CHANGES_REQUESTED** [review]
Thank you very much! I just left a few comments.

File: examples/clustering/plot_dtw_boss.py (hunk 2 above)
Comment (lucasplagwitz) [reply examples/clustering/plot_dtw_boss.py]: Couldn't see exactly how the error shows up in my browser. I have simply included your suggestion. Please check again if your problem is solved this way.

Comment (johannfaouzi) [reply examples/clustering/plot_dtw_boss.py]: Capture d’écran 2022-06-16 à 10 50 01

I see a line break, but I don't update my browser very often, so it may explain this. It's not really important anyway, thanks for trying out!
